### ISSUE TITLE
split ACME certificate storage in consul to allow more than 100 certificates 

### ISSUE THREAD

**Issue by dazoot** [containous/traefik#3833]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

Instad of storing all ACME certificates in a compressed consul `k/v`, would be nice to have an `k/v` entry per frontend. This way the 512Kb limit per value of `consul` won't be reached and there won't be a max 100 certificates limitation for `traefik`.

**Comment by shane-axiom** [comment]
Yes, this is an adoption blocker for us as well!

**Comment by jrisebor** [comment]
This is also blocking us from adopting consul.

The majority of our certs are on a single frontend, so breaking up the k/v entries by frontend may not be the best way to divide them. I think one k/v pair per cert could make more sense (especially for SAN certs).

**Comment by 2union** [comment]
`*****************************************
*******************+z++******************
*****************#zxxnnz+****************
****************+MxWWWWMxz***************
***************+xWWWWWWWWx+**************
***************nWWMMxnnxxWz**************
**************+WWMz#++++#xM**************
+*************zWMn#++**++nW+*************
**************zWMz#++++++#W#*************
+*************+Wnzz#++#zz#n+*************
++++++********#xnznxn+zxz+#+*************
+++++++++*****+nzzzzz###++#+*************
++++++++++++**+zz#+#z#++++#**************
++++++++++++++++nz##z#++++z**************
++++++++++++++++zn#zxzz####+++++*+****++*
+++++++++++++++++xzxWMMMzn++++**++****++*
+++++++++++++++++Wxxnnnzxn++++++++**++++*
++++++++++++++++xWWxnxzzM#+++++++++*++++*
++++++++++++++zxWxMWxMnMx#+xn++++++++++++
+++++++++++++xMW@xnWWWWMz#+MWx#++++++++++
+++++++++++#WWW@@Mnnxnnzz#xWWWMz+++++++++
++++++++++zMW@W@@Mnnzzzz##WWWWWMz++++++++
+++++++++nMMWWWWWMzzzzz#+xWWWWWMMz+++++++
++++++++zWWWW@WWWWx####+nWWWWWWWMx#++++++
+++++++#MWWWW@WWWWWMzznMWWWWWWWWMMn++++++
+++++++xWWWWW@WWWWWWWWWWWWWWWWWWMWMz+++++
++++++nWWWWWWWWWWWWWWWWWWWWWWWWWWWMM+++++
+++++zMWWWWWWWWWWWWWWWWWWWWWWWWWWWWMn++++
++++#MWWWWWW@WWWWWWWWWWWWWWWWWWWWWWMM++++
++++xWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWW#+++
++++#xWWWWW@WWWWWWWWWWWWWWWWWWW@WWWWWx+++
+++++zMWWWW@WWWWWWWWWWWWWWWWWWW@WWWWWz+++
+++++nxnxWW@WWWWWWWWWWWWWWWWWWW@WWWMn++++
+++++znz#+nW@WWWWWWWWWWWWWWWWWW@WWnzn#+++
+++++zzz#++M@WWWWWWWWWWWWWWWWWW@M###zz+++
+++++zz#++#@@WWWWWWWWWWWWWWWWWWWn###zz+++
++++#nz#+#n@@WWWWWWWWWWWWWWWWWWxz###zz+++
++++#nz###x@@WWWWWWWWWWWWWWWWWWn#z###z#++
++++zz###zn@@WWWWWWWWWWWWWWWWWW#*nzzzz#++
++++zz##z#n@@WWWWWWWWWWWWWWWWMM+*zzzzz#++
++++nz#zz+x@@WWWWWWWWWWWWWWWWMM+*+nzzzz+*
++++nzzzz+W@@WWWWWWWWWWWWWWWWWM#**nnnzz#*
++++nzzzz#W@@WWWWWWWWWWWWWWWWWMz**znzzz++
++++nzzzzn@@@WWWWWWWWWWWWWWWWWMn**nnzzz++
++++nnzzzx@@@@WWWWWWWWWWWWWWWMWn*+nnzzz#*
++++znzzzx@@@@WWWWWWWWWWWWWWWW@n++nzzzz#*
++++zzzzzzW@@@@WWWWWWWWWWWWWWW@W++nzzzz#*
++++#zzzznW@@@@WWWWWWWWWWWWWW@@@++nnzzz++
+++++nzzznW@@@@@WWWWWWWWWWWWW@@W++nnzzz++
+++++zzzzzM@@@@@@WWWWWWWWWWW@@@W#+nnzzz++
++++++nzzzxW@@@@@@@WWWWWWW@@@@@@z+nnzz#++
++++++zzzznWW@@@@@@@@WWWW@@M@@@W#+n#zz+++
+++++++zzzzMW@@@@MMWW@@WWWWnxWWWznWWnz+++
+++++++zzzznMxxxn#W@@@WWWW@Wz###zznMxx+++
+++++++nzzzzzz##zx@@@@@@@@@@@Mzzzz#nMM+++
+++++++Mnz##zz#xW@@@@@@@@@@@@@xz####nz+++
++++++zWMzzz##zWWWW@@@@@@@@@Wn#####z#++++
++++++xWWnzz###nWWW@@@@@@@@M#znz###z+++++
+++++#MWWxzzz#zzzMW@@@@@@@x#xMzzz###+++++
+++++nWWWMnz#z#xnzM@@@@@@@WWM#nnzzz++++++
+++++WWWWWM#nzn#MMz@@@@@@@@Wzxz#MzMz+++++
++++nWWWWWWxznzn#WW@@@@@@@@WWnzWzxMx+++++
++++WW@WWWWWzxnxxn@@@@@@@@WWWMWxxWWM#++++
++++W@@WWWWWMxWnWM@@@@@@@WWWWWWWWWWWn++++
+++zW@@WWWWWWWWWW@@@@@@@@WWWWWWWWWWMM++++
+++xWWWWWWWWWWW@@@Wz##nWWWWWWWWWWWWWM#+++
+++WWWWWWWWWWWWWWz+++++#M@WWWWWWWWWWMz+++
++zWWWWWWWWWWWWM#++++++++zWWWWWWWWWWMn+++
++nWWWWWWWWWWWM+++++++++++nWWWWWWWMMMx+++
++xWWWWWWWWWWM+++++++++++++MWWWWWWMMMM+++
++MWWWWWWWWWW+++++++++++++++WWWWWWMMMM#++
++MWWWWWWWWWn+++++++++++++++zWWWWWWMMMz++
`

**Comment by samuelhautcoeur** [comment]
We're blocked as well with this lmitation.

**Comment by dazoot** [comment]
Any plans to include this feature request in the upcoming 2.0 release ? CC @ldez ?

**Comment by wido** [comment]
I would like to chime in here. Is somebody working on this? If not, how can somebody help or sponsor this?

**Comment by 2union** [comment]
> I would like to chime in here. Is somebody working on this? If not, how can somebody help or sponsor this?

You can use Zookeeper as KV storage for traefik, it's allows add more then 100 ACME certs.

**Comment by wido** [comment]
@2union Any ideas how many certificates you can safely load? I'm looking at about 100k certs.

**Comment by sarkistlt** [comment]
@wido I think it won't work for you, have the same problem.
@2union using Zookeeper vs consul doesn't solve the problem, you can store 100 certs with any KV or just a JSON, point is, currently it just stores all certs under one key, so imagine having 100k certs as a value of one key, what will be the latency or if there any glitch during writing the value, then all 100k will be affected. The best would be if traefik would store certs per domain.

**Comment by wido** [comment]
Yes, so that's the main issue @sarkistlt , everything is stored in a single key in the KV store.

I also found that Consul has a hard time with listing many keys and when Traefik is trying to watch the config store. So I doubt that it will work properly with so many keys.

**Comment by reverseppp** [comment]
@sarkistlt hey i saw you in another issue : datastore sync error expected .... got .... with traefik coupled with consul; Can you please tell me how did you solve your issue, i am having the same problem and i can't go to production because of that 
Thanks

**Comment by sarkistlt** [comment]
@reverseppp hey, can you please point to the issue you are talking about?

**Comment by reverseppp** [comment]
@sarkistlt thank you i've finally solved the problem
Regarding this issue, i read that the certificates are compressed before being put in the kv store; So can i store all my 100+ certificates in consul ? or should i switch to etcd or zookeeper ? Because i have about 150 domains behind traefik, and each have its own certificates

**Comment by flokli** [comment]
@reverseppp No matter whether certs are compressed or not, storing all of them behind one single key will hit some limits sooner or later.

@wido you mean, splitting ACME storage into one key per cert would be problematic, as having consul watch a lot of keys is a problem for consul? I was hoping the [keyprefix](https://www.consul.io/docs/agent/watches.html#type-keyprefix) watch type could easily handle watching a tree with some 100k-ish keys underneath, obviously we don't want to register 100k watchers manually…

**Comment by wido** [comment]
@flokli Yes, that is what I meant. Because 100 certs is a limitation right now.

I think I know what you mean with watching so many keys on Consul. I tried consul as a configuration backend and with many frontends and backends (10k) it would kill Consul. 100% CPU and just very slow.

**Comment by sarkistlt** [comment]
@wido hm, haven't tried with 10k frontends yet, but will be close to hit that number, so what the alternative you found to store your configs?

@reverseppp currently the best approach I have is to not use traefik for ACME. Write you own script that will generate and watch for renewal, and you can store it for example as an secret in k8s, or as a files on SSD

**Comment by wido** [comment]
@sarkistlt File backend. We now use a process to generate this file with Python and have Traefik watch it.

**Comment by michiels** [comment]
@wido Interesting. We are also still using the file backend on a single server that simply runs Traefik to reverse-proxy custom domains to our backend applications.

Do you already have some HA availability setup with the File storage that you can share your thoughts on? We'd like to make our Traefik setup more rebust and have the ability to reboot it's server without downtime, so essentially make the ACME.json shared across Traefik workers.

**Comment by wido** [comment]
This goes rather off-topic here and no longer applies to the certs in the KV store.

Short story: We use Salt to provision the Traefik servers with Files containing their routes.

**Comment by X4mp** [comment]
Hey, This ~100 cert limitation is also a blocker for us to migrate our managed domains to traefik with consul backend. Is there currently anyone working on it?

**Comment by shane-axiom** [comment]
@ldez @nmengin Would love to hear some input from devs on this. The current [acme docs](https://github.com/containous/traefik/blob/master/docs/content/https/acme.md#in-a-file) say that a KV store must be used instead of file storage for multiple Traefik nodes:

> For concurrency reason, this file cannot be shared across multiple instances of Traefik. Use a key value store entry instead.

However as discussed here this isn't feasible beyond small/test deployments, and it's easy to quickly overwhelm this limitation if you're doing branch deployments in CI, etc. AFAICT this limitation isn't changing in Traefik 2.0.

Are devs considering a fix for this issue? Too busy but will revisit later? WONTFIX?

**Comment by AndrewSav** [comment]
I think that this is going to be supported in https://containo.us/traefikee/ if you are willing to pay for it. "Distributed Let's Encrypt" is mentioned as one of the features.

**Comment by shane-axiom** [comment]
Totally understandable if that’s the case, but some clarity would be appreciated

**Comment by X4mp** [comment]
> I think that this is going to be supported in https://containo.us/traefikee/ if you are willing to pay for it. "Distributed Let's Encrypt" is mentioned as one of the features.

Is this "really" what we were requesting here? Can one of the devs comment on what "Distributed Let's Encrypt" mean? Some clarity would indeed be appreciated.

**Comment by AndrewSav** [comment]
I posted a question [here](https://community.containo.us/t/please-explain-distributed-lets-encrypt-feature-of-traefikee/872); let's hope we get some love ;)

**Comment by nmengin** [comment]
Hello,

This enhancement is linked to a feature that no longer exists in the supported versions.
I close it.

We'll re-open it later if necessary.
